### PR TITLE
HIVE-24259: [CachedStore] Optimise get all constraint api

### PR DESCRIPTION
**Pull request by ashish-kumar-sharma**
### What changes were proposed in this pull request?

1. Redundant check if table is present or not
2. Currently in order to get all constraint form the cachedstore. 6 different call is made with in the cached store. Which led to 6 different call to raw store

### Why are the changes needed?

1. Check only once if table exit in cached store.
2. Instead of calling individual constraint in cached store. Add a method which return all constraint at once and if data is not consistent then fall back to rawstore.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

running all UTs

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java [hunk 1]
```diff
@@ -2836,14 +2836,32 @@ long getPartsFound() {
...
+    Table tbl = sharedCache.getTableFromCache(catName, dbName, tblName);
+    if (tbl == null) {
+      // The table containing the constraints is not yet loaded in cache
+      return rawStore.getAllTableConstraints(catName, dbName, tblName);
+    }
+    SQLAllTableConstraints constraints = sharedCache.listCachedAllTableConstraints(catName, dbName, tblName);
+
+    // if any of the constraint value is missing then there might be the case of partial constraints are stored in cached.
+    // So fall back to raw store for correct values
+    if (constraints != null && CollectionUtils.isNotEmpty(constraints.getPrimaryKeys()) && CollectionUtils
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2856]: Majority of the calls are likely to be redirected to RawStore and lose the advantage of cache. How about an optional flag in SQLAllTableConstraints and TableWrapper to mark if it is a complete snapshot of constraints?

Comment (adesh-rao) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2856]: +1.

Comment (adesh-rao) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2856]: Also, what if the table just a primary keys and not other constraints?

nonEmpty(otherConstraints) will return true and we will not cache anything?

Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2856]: Good idea. I will check how we can implement a flag.

Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2821]: Adding a flag required bigger change. So for now I am reducing the scope of this PR to optimise following
1. Check only once if table exit in cached store.
2. Instead of calling individual constraint in cached store. Add a method which return all constraint at once and if data is not consistent then fall back to rawstore.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java [hunk 2]
```diff
@@ -2844,15 +2814,10 @@ public SQLAllTableConstraints getAllTableConstraints(String catName, String dbNa
...
-    Table tbl = sharedCache.getTableFromCache(catName, dbName, tblName);
-    if (tbl == null) {
-      // The table containing the constraints is not yet loaded in cache
-      return rawStore.getAllTableConstraints(catName, dbName, tblName);
-    }
     SQLAllTableConstraints constraints = sharedCache.listCachedAllTableConstraints(catName, dbName, tblName);
 
-    // if any of the constraint value is missing then there might be the case of partial constraints are stored in cached.
-    // So fall back to raw store for correct values
+    /* If constraint value is missing then there might be the case that table is not stored in cached or
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2819]: Add a TODO that this check is inefficient as all calls will hit rawstore if even one of the constraint type is not set for the given table.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 1 above)
Comment (sankarh) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2821]: Why is it hard to set a flag for consistency? If we hit else flow in this case, we shall try to update the cache and if that fails for some reason, then set the flag to false else set to true. Same with pre-warm and refresh flows.
Also, we need to check only the flag to validate the consistency of cache. No need to check for null or empty constraints.

**Review by sankarh — CHANGES_REQUESTED** [review]

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 1 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java]: Added flag

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java [hunk 3]
```diff
@@ -2397,7 +2397,7 @@ public SQLAllTableConstraints listCachedAllTableConstraints(String catName, Stri
 
   public List<SQLForeignKey> listCachedForeignKeys(String catName, String foreignDbName, String foreignTblName,
                                                    String parentDbName, String parentTblName) {
-    List<SQLForeignKey> keys = new ArrayList<>();
+    List<SQLForeignKey> keys = null;
```
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:2540]: done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 2 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java]: done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java [hunk 4]
```diff
@@ -309,12 +302,19 @@ public int getObjectSize(Class<?> clazz, Object obj) {
     private Map<String, SQLDefaultConstraint> defaultConstraintCache = new ConcurrentHashMap<>();
     private Map<String, SQLCheckConstraint> checkConstraintCache = new ConcurrentHashMap<>();
 
+    private boolean pk = false;
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:305]: Using too many flags is complex. Check if we can simplify it using below logic. 
1) Fetch all constraints and update cache -> If this update is successful, then set "isAllConstraintsSet=true"
2) If partially successful due to memory constraint, then set "isAllConstraintsSet=false" and set bit position of completed constraint "long constraintStatusBits". We can assign one bit per constraintType and generally in the order in which we update it. It means, if we encounter a bit which is unset, then rest of the bits are unset too.
3) We never refresh individual constraints. Always try to update all constraints by using same method which follow above logic to set the flag. Shall remove all individual update(CachedStore.updateTablePrimaryKeys...) and refresh(SharedCache.refreshPrimaryKeysInCache...) methods for each types. This ensure, we always cache a valid snapshot.
4) When read any constraint, first check isAllConstraintsSet==true. If true, then assume all are valid and return the cached entry immediately. If false, then just the corresponding bit in "constraintStatusBits" and if true, then return the cached entry or else return null and invoke RawStore.

Note: We can also get rid of isAllConstraintsSet flag by using constraintStatusBits = -1 for "true" scenario.

**Review by sankarh — CHANGES_REQUESTED** [review]

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java (hunk 4 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java]: Changes are as follows 
1. Replace 6 flag with 1 flag
2. In cache refresh/update flow removed 6 call to raw metastore with getAllTableConstraints()
3. While adding table to cache via createTable and createTableWithConstraint isConstraintValid flag is set as true.
4. If  isConstraintValid flag is true then return data from cache else redirect to raw metastore.
5. While adding individual constraint to metastore if  isConstraintValid is true that means constraints snapshot is valid and new constraint can be updated in cachestore else skip update.
6. isConstraintValid will only be true if all the constraint value are valid in the metastore else redirect to raw metastore (For the case of partial constraint storage due to memory shortage)

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java [hunk 5]
```diff
@@ -1325,6 +1204,7 @@ private void validateTableType(Table tbl) {
     }
     validateTableType(tbl);
     sharedCache.addTableToCache(catName, dbName, tblName, tbl);
+    sharedCache.addTableConstraintsToCache(catName,dbName,tblName,new SQLAllTableConstraints());
```
Comment (adesh-rao) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:1207]: There is a separate api for createTableWithConstraints. 

Is this api also used when table is created with constraints?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java [hunk 6]
```diff
@@ -418,12 +424,12 @@ private void updateMemberSize(MemberName mn, Integer size, SizeMode mode) {
       }
 
       switch (mode) {
-        case Delta:
-          this.memberObjectsSize[mn.ordinal()] += size;
-          break;
-        case Snapshot:
-          this.memberObjectsSize[mn.ordinal()] = size;
-          break;
+      case Delta:
```
Comment (adesh-rao) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:427]: nit: need extra spaces before case?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java [hunk 7]
```diff
@@ -2814,21 +2649,12 @@ long getPartsFound() {
     catName = StringUtils.normalizeIdentifier(catName);
     dbName = StringUtils.normalizeIdentifier(dbName);
     tblName = StringUtils.normalizeIdentifier(tblName);
-    if (!shouldCacheTable(catName, dbName, tblName) || (canUseEvents && rawStore.isActiveTransaction())) {
+    if (!shouldCacheTable(catName, dbName, tblName) || (canUseEvents && rawStore.isActiveTransaction()) || !sharedCache
+        .isTableConstraintValid(catName, dbName, tblName)) {
       return rawStore.getCheckConstraints(catName, dbName, tblName);
     }
+    return sharedCache.listCachedCheckConstraint(catName, dbName, tblName);
```
Comment (adesh-rao) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2656]: below check was present earlier:

```
if (CollectionUtils.isEmpty(keys)) {	
      return rawStore.getCheckConstraints(catName, dbName, tblName);	
}
```

Why is this removed? Is this check moved somewhere else?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 5 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:1207]: This line was added to update table cache with all constraints null and flag set to true. So avoid redirection to raw store.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 7 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2656]: Added flag - isConstraintValid to check null value is consistent instead of going to raw store.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java [hunk 8]
```diff
@@ -715,154 +747,105 @@ public void removeConstraint(String name) {
...
-      Map<String, SQLForeignKey> newKeys = new ConcurrentHashMap<>();
-      try {
-        tableLock.writeLock().lock();
-        int size = 0;
-        for (SQLForeignKey key : keys) {
-          if (compareAndSetMemberCacheUpdated(MemberName.FOREIGN_KEY_CACHE, true, false)) {
-            LOG.debug("Skipping foreign key cache update for table: " + getTable().getTableName()
-                    + "; the foreign keys are already refreshed.");
-            return;
+    private boolean refreshConstraint(List constraints, MemberName mn) {
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:765]: This also always returns true and in tune make isConstraintsValid=true always.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java [hunk 9]
```diff
@@ -1788,47 +1771,16 @@ public boolean populateTableInCache(Table table, TableCacheObjects cacheObjects)
...
-        return false;
-      }
-    }
-    tblWrapper.setMemberCacheUpdated(MemberName.DEFAULT_CONSTRAINT_CACHE, false);
-
-    if (CollectionUtils.isNotEmpty(constraints.getCheckConstraints())) {
-      if (!tblWrapper.cacheCheckConstraints(constraints.getCheckConstraints(), true)) {
-        return false;
-      }
+    if(tblWrapper.cacheConstraints(constraints,true)) {
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:1774]: nit: Space before  ( and after comma

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java [hunk 10]
```diff
@@ -715,154 +747,105 @@ public void removeConstraint(String name) {
...
+          }
+        }
+        checkConstraintCache = newCc;
+        break;
+      default:
+        LOG.error("Should not reach here");
+        break;
+      }
+      updateMemberSize(mn, size, SizeMode.Snapshot);
+      setMemberCacheUpdated(mn, false);
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:845]: Why do we set this flag to false?

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java [hunk 11]
```diff
@@ -2187,7 +2139,7 @@ public void addPrimaryKeysToCache(String catName, String dbName, String tblName,
       cacheLock.readLock().lock();
       String tblKey = CacheUtils.buildTableKey(catName, dbName, tblName);
       TableWrapper tblWrapper = tableCache.getIfPresent(tblKey);
-      if (tblWrapper != null) {
+      if (tblWrapper != null && tblWrapper.isConstraintsValid()) {
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:2142]: Keep tblWrapper != null within ().

Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:2142]: Why do we update cache for each constraint individually? We should always cache a single snapshot of constraints of table. If not, isConstraintsValid flag wouldn't make sense.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sankarh) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:2352]: Still keys can be referenced with null at line 2365.

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java [hunk 12]
```diff
@@ -1558,6 +1558,7 @@ public void testPrimaryKeys() {
     Table tbl = createUnpartitionedTableObject(db);
 
     sharedCache.addTableToCache(DEFAULT_CATALOG_NAME, "db", tbl.getTableName(), tbl);
+    sharedCache.addTableConstraintsToCache(DEFAULT_CATALOG_NAME,"db", tbl.getTableName(), new SQLAllTableConstraints());
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java:1561]: nit: Need space after comma

**Review by sankarh — CHANGES_REQUESTED** [review]

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java (hunk 11 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:2142]: Q Why do we update cache for each constraint individually? 

Ans - To update the cache for specific constraint for query like "ALTER TABLE table_name ADD CONSTRAINT" instead of invalidating the whole cache and updating again because that will lead to cache miss or extra network call is required to delete and update all cache constraints.

Q We should always cache a single snapshot of constraints of table. If not, isConstraintsValid flag wouldn't make sense.

Ans - So if isConstraintsValid is valid that means there is valid copy of constraint present then we can add the incoming constraint and save redundant call. if isConstraintsValid is false then skip because cached constraint are inaccessible and waiting for refreshed

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java]: done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java (hunk 9 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java]: done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java (hunk 11 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java]: done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java (hunk 10 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java]: because it is a snapshot

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java (hunk 12 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/cache/TestCachedStore.java:1561]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java (hunk 8 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java:786]: Yes

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java [hunk 13]
```diff
@@ -506,10 +512,18 @@ boolean cacheCheckConstraints(List<SQLCheckConstraint> checkConstraints, boolean
...
+              && cacheConstraints(constraints.getNotNullConstraints(), fromPrewarm, MemberName.NOTNULL_CONSTRAINT_CACHE)
+              && cacheConstraints(constraints.getCheckConstraints(), fromPrewarm, MemberName.CHECK_CONSTRAINT_CACHE);
+      return this.isConstraintsValid;
+    }
+
     // Common method to cache constraints
-    private boolean cacheConstraints(List constraintsList,
-                             boolean fromPrewarm,
-                             MemberName mn) {
+    private boolean cacheConstraints(List constraintsList, boolean fromPrewarm, MemberName mn) {
```
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java]: In case of wrong member type.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/events/CreateTableEvent.java [hunk 14]
```diff
@@ -29,11 +30,14 @@
 
   private final Table table;
   private final boolean isReplicated;
+  private final SQLAllTableConstraints constraints;
 
-  public CreateTableEvent(Table table, boolean status, IHMSHandler handler, boolean isReplicated) {
+  public CreateTableEvent(Table table, boolean status, IHMSHandler handler, boolean isReplicated,
+      SQLAllTableConstraints constraints) {
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/events/CreateTableEvent.java:36]: We have separate events for add individual constraints. I think, we need not keep it as part of create table event. 
It adds additional overhead when transmit events for incremental replication. Also, takes additional storage in HMS.

**Review by sankarh — CHANGES_REQUESTED** [review]

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/events/CreateTableEvent.java (hunk 14 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/events/CreateTableEvent.java]: Removed

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java [hunk 15]
```diff
@@ -2187,8 +2160,10 @@ public void addPrimaryKeysToCache(String catName, String dbName, String tblName,
       cacheLock.readLock().lock();
       String tblKey = CacheUtils.buildTableKey(catName, dbName, tblName);
       TableWrapper tblWrapper = tableCache.getIfPresent(tblKey);
-      if (tblWrapper != null) {
-        tblWrapper.cachePrimaryKeys(keys, false);
+      if ((tblWrapper != null) && tblWrapper.isConstraintsValid()) {
+        // Because lake of snapshot freshness validation.
+        // For now disabling cached constraint snapshot addition in parts by invalidating constraint snapshot.
+        tblWrapper.inValidateConstraints();
```
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/SharedCache.java]: Resolved

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java [hunk 16]
```diff
@@ -2674,28 +2554,19 @@ long getPartsFound() {
     catName = StringUtils.normalizeIdentifier(catName);
     dbName = StringUtils.normalizeIdentifier(dbName);
     tblName = StringUtils.normalizeIdentifier(tblName);
-    if (!shouldCacheTable(catName, dbName, tblName) || (canUseEvents && rawStore.isActiveTransaction())) {
+    if (!shouldCacheTable(catName, dbName, tblName) || (canUseEvents && rawStore.isActiveTransaction()) || !sharedCache
```
[the remark this answers is not in the document]
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java]: Done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java [hunk 17]
```diff
@@ -144,4 +147,15 @@ public static boolean matches(String name, String pattern) {
     }
     return false;
   }
+
+  public static boolean validateShouldCacheTableCanUseEventIsActiveTransaction(String catName, String dbName,
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java:151]: It is not used anywhere other than below method. We can remove it and move these validations there.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java [hunk 18]
```diff
@@ -144,4 +147,15 @@ public static boolean matches(String name, String pattern) {
...
+      String tblName, boolean canUseEvents, RawStore rawStore) {
+    return !shouldCacheTable(catName, dbName, tblName) || (canUseEvents && rawStore.isActiveTransaction());
+  }
+
+  public static boolean validateShouldCacheTableCanUseEventIsActiveTransactionIsConstraintsValid(String catName,
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java:156]: The method name is too long and not readable. Shall make it shouldGetConstraintFromRawStore().

Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java:156]: Shall move this method as private method inside CachedStore to avoid passing too many arguments.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java [hunk 19]
```diff
@@ -2554,8 +2554,7 @@ long getPartsFound() {
     catName = StringUtils.normalizeIdentifier(catName);
     dbName = StringUtils.normalizeIdentifier(dbName);
     tblName = StringUtils.normalizeIdentifier(tblName);
-    if (!shouldCacheTable(catName, dbName, tblName) || (canUseEvents && rawStore.isActiveTransaction()) || !sharedCache
-        .isTableConstraintValid(catName, dbName, tblName)) {
+    if (validateShouldCacheTableCanUseEventIsActiveTransactionIsConstraintsValid(catName,dbName,tblName,canUseEvents,rawStore,sharedCache)) {
```
Comment (sankarh) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java:2557]: nit: Add space after each parameter.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java (hunk 17 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java]: removed

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java (hunk 18 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java]: done

Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CacheUtils.java]: done

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java (hunk 19 above)
Comment (ashish-kumar-sharma) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/cache/CachedStore.java]: done

**Review by sankarh — APPROVED** [review]

**Review by adesh-rao — APPROVED** [review]
